source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h

```diff
@@ -15,20 +18,27 @@ namespace LocalRateLimit {
 
 class LocalRateLimiterImpl {
 public:
-  LocalRateLimiterImpl(const std::chrono::milliseconds fill_interval, const uint32_t max_tokens,
-                       const uint32_t tokens_per_fill, Event::Dispatcher& dispatcher);
+  LocalRateLimiterImpl(
+      const std::chrono::milliseconds fill_interval, const uint32_t max_tokens,
+      const uint32_t tokens_per_fill, Event::Dispatcher& dispatcher,
+      const Protobuf::RepeatedPtrField<
+          envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>& descriptors);
   ~LocalRateLimiterImpl();
 
-  bool requestAllowed() const;
+  bool requestAllowed(const std::vector<RateLimit::LocalDescriptor>& request_descriptors) const;
 
 private:
   void onFillTimer();
+  void onFillTimerHelper(const RateLimit::TokenState& state, const RateLimit::TokenBucket& bucket);
+  void onFillTimerDescriptorHelper();
+  bool requestAllowedHelper(const RateLimit::TokenState& tokens) const;
 
-  const std::chrono::milliseconds fill_interval_;
-  const uint32_t max_tokens_;
-  const uint32_t tokens_per_fill_;
+  RateLimit::TokenBucket token_bucket_;
   const Event::TimerPtr fill_timer_;
-  mutable std::atomic<uint32_t> tokens_;
+  TimeSource& time_source_;
+  RateLimit::TokenState tokens_;
+  absl::flat_hash_set<RateLimit::LocalDescriptor> descriptors_;
```

Comment:
FYI: matching is now done with a single hash set lookup.

Reply:
I think now we are doing order match in the order of rate limit actions specified in the route. We should specify this in the documentation....

Reply:
Agree, it's the first matching descriptor that is activating rate limit. I wonder if we want _all_ matches?

Reply:
I think the first match more sense to control runtime costs. We could probably switch to a trie later on.

Reply:
IMO first match is OK, but please clearly document it. I think eventually we could add config for all, etc.

Reply:
The only corner case I am concerned about is this...
```yaml
route_config:
    name: local_route
    virtual_hosts:
    - name: local_service
      domains: ["*"]
      routes:
      - match: { prefix: "/foo" }
        route: { cluster: service_protected_by_rate_limit }
        typed_per_filter_config:
          envoy.filters.http.local_ratelimit:
            "@type": type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
            stat_prefix: test
            token_bucket:
              max_tokens: 1000
              tokens_per_fill: 1000
              fill_interval: 60s
            filter_enabled:
              runtime_key: test_enabled
              default_value:
                numerator: 100
                denominator: HUNDRED
            filter_enforced:
              runtime_key: test_enforced
              default_value:
                numerator: 100
                denominator: HUNDRED
            response_headers_to_add:
              - append: false
                header:
                  key: x-test-rate-limit
                  value: 'true'
            descriptors:
            - entries:
              - key: client_id
                value: foo
              - key: path
                value: /foo/bar
              token_bucket:
                max_tokens: 10
                tokens_per_fill: 10
                fill_interval: 60s
            - entries:
              - key: path
                value: /foo/bar
              token_bucket:
                max_tokens: 1000
                tokens_per_fill: 1000
                fill_interval: 60s
      - match: { prefix: "/" }
        route: { cluster: default_service }
      rate_limits:
      - actions: # any actions in here
        - request_headers:
            header_name: ":path"
      - actions: # any actions in here
        - generic_key:
            descriptor_value: "foo"
            descriptor_key: "client_id"
        - request_headers:
            header_name: ":path"
```

Here user specified the ordering that {client_id:foo,path:/foo/bar} as more restrictive but because rate limit actions ordering decide the order of evaluation, {path:/foo/bar} will be selected to decide the rate limiting...
just something we should document IMO...

Reply:
I think the order of actions is important is both global and local descriptors, so it's not much different from global.